### PR TITLE
WIP: First Cut - this doesn't work yet - a work-in-progress.

### PR DESCRIPTION
**Pull request by mbeckerle**
Doesn't work in Daffodil for lack of "V" textNumberPattern feature.

Doesn't work in IBM DFDL for lack of binaryCalendarRep="packed" support. 
Also needs dfdl:calendarTimeZone="UTC" because "" is not accepted - this will require changing the test logical data to have "+00:00" on the dates.

### CODE REVIEW

**Comment by mbeckerle** [comment]
Ok, to be clear, IBM DFDL has lots of working Cobol capability. What isn't working is my little example here yet. Presumably it can be made to work with a workaround to the issues blocking me.

File: src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+          binaryFloatRep="ieee" 
+          binaryNumberCheckPolicy="strict" 
+          binaryPackedSignCodes="C D F C" 
+          byteOrder="bigEndian" 
+          calendarCenturyStart="53" 
+          calendarCheckPolicy="strict" 
+          calendarFirstDayOfWeek="Monday"           
+          calendarLanguage="en-US" 
+          calendarObserveDST="yes" 
+          calendarTimeZone=""
```
Comment (mbeckerle) [inline src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd:38]: IBM portability issue - needs to be a real time zone e.g., "UTC".

File: src/test/resources/com/tresys/cobol1/tests.tdml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  <tdml:unparserTestCase name="test1" model="com/tresys/cobol1/xsd/cobol1.dfdl.xsd">
+    <tdml:document>
+      <tdml:documentPart type="text" encoding="ebcdic-cp-us">1234567890123456789</tdml:documentPart>
+      <tdml:documentPart type="byte">0020161C</tdml:documentPart>
+      <tdml:documentPart type="byte">0020161C</tdml:documentPart>
+      <tdml:documentPart type="byte">0020161C</tdml:documentPart>
+      <tdml:documentPart type="byte">0020161C</tdml:documentPart>
+      <tdml:documentPart type="byte">99999C</tdml:documentPart>
+      <tdml:documentPart type="byte">270f</tdml:documentPart>
```
Comment (mbeckerle) [inline src/test/resources/com/tresys/cobol1/tests.tdml:30]: This binary number may have an alignment to a 2-byte boundary - depends on the cobol implementation. Should set this up for typical IBM s390 Cobol. 

The same is true for the other binary numbers in the data (could be 2, 4, or 8 byte boundaries).

File: src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    dfdl:representation="binary"
+    dfdl:binaryNumberRep="binary"
+    dfdl:length="2"> 
+    <xs:restriction base="xs:int">
+      <xs:maxInclusive value="9999"/>
+      <xs:minInclusive value="-9999"/>
+    </xs:restriction>
+  </xs:simpleType>  
+  
+  <xs:simpleType name="binary5Digits"
```
Comment (mbeckerle) [inline src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd:209]: Is there any possibility of a 3-byte binary integer?

File: src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        <xs:element name="INTERMEDIARY-CTRL-NUM" type="xs:string" dfdl:length="23" />
+      </xs:sequence>
+    </xs:complexType>
+  </xs:element>
+  
+  <!-- 
+    IBM DFDL doesn't support binaryCalendarRep="packed" so we have to 
+    hack around this using BCD, and surrounding things that create the 
+    leading 0 nibble, and trailing "C" nibble which would exist for packed.
+    -->
```
Comment (mbeckerle) [inline src/main/resources/com/tresys/cobol1/xsd/cobol1.dfdl.xsd:134]: This workaround isn't working-around yet. There may be a better way to do this.

**Review by mbeckerle — COMMENTED** [review]
Not yet ready for prime time.

**Comment by smhdfdl** [comment]
@mbeckerle When IBM creates DFDL schema for COBOL, we auto-generate them from the COBOL copybook or program. There is no concept of a date/time data type in COBOL, and that kind of data normally would be represented as string or zoned decimal. We have never needed to generate binaryCalendarRep="packed".

**Comment by smhdfdl** [comment]
@mbeckerle Suggested default properties for COBOL ... as there is no date/time concept in COBOL, there is nothing declared for date/time.

```
<dfdl:format 
  representation="text" 
  escapeSchemeRef="" 
  ignoreCase="yes" 
  initiatedContent="no"
  alignment="1" alignmentUnits="bytes" 
  textStandardBase="10"  textNumberCheckPolicy="lax"  
  textStandardZeroRep="0" textStandardExponentCharacter="E"
  textStandardDecimalSeparator="." textStandardGroupingSeparator="," 
  textNumberRoundingMode="roundUp"  textNumberRounding="pattern"
  textNumberJustification="right" textPadKind="padChar" textTrimKind="padChar" 
  sequenceKind="ordered"  floating="no"
  textOutputMinLength="0"
  truncateSpecifiedLengthString="no"
  nilValueDelimiterPolicy="both" emptyValueDelimiterPolicy="none"
  documentFinalTerminatorCanBeMissing="no"
  useNilForDefault="yes" 
  choiceLengthKind="explicit" 
  fillByte="%#r00;"
  binaryNumberCheckPolicy="lax" 
  binaryBooleanTrueRep="1" binaryBooleanFalseRep="0"
  textBooleanTrueRep="true" textBooleanFalseRep="false" 
  textBooleanJustification="left" textBooleanPadCharacter="%SP;" 
  lengthKind="explicit" lengthUnits="bytes" 
  binaryNumberRep="binary"
  binaryPackedSignCodes="C D F C"
  outputNewLine="%CR;%LF;" textBidi="no" 
  initiator="" terminator="" 
  separator=""  separatorPolicy="required" separatorPosition="infix" 
  leadingSkip="0" trailingSkip="0" 
/>
```

**Comment by smhdfdl** [comment]
@mbeckerle Consider declaring simple types for all the COBOL PIC types. You can carry some properties on the simple types to avoid duplication. It also makes sense to a someone familiar with COBOL. Example:
```
	<xsd:simpleType name="PIC9-Comp3__short" dfdl:representation="binary"
		dfdl:binaryNumberRep="packed">
		<xsd:restriction base="xsd:unsignedShort">
			<xsd:minInclusive value="0" />
			<xsd:maxInclusive value="9999" />
		</xsd:restriction>
	</xsd:simpleType>
```
I can see if IBM can donate the full list of simple types.

**Comment by mbeckerle** [comment]
This got closed by accident. Should be reopened.

**Pull request merged**